PEP 735: Draft 2 of Dependency Groups in pyproject.toml (#3566)
python/peps: peps/pep-0735.rst

Reviewer: The spec has a hole where it defines "include" objects, but doesn't mention dependency specifiers (aka PEP 508 specifiers) as acceptable input. The opening paragraph of the Specification section hints at it (by mentioning "requirements specifiers"), but actually never defines it.

Author: Either I missed this in the first draft or I got a little *too enthusiastic* about making things shorter.

Either way, I'll rectify in a new revision.

Author: Actually, the current content says this, but perhaps not with the best phrasing (first section under Specification).
I'll try to rephrase to be shorter and more direct, so that this is clarified.

Author: I've tried to address this and also be more consistent about calling TOML tables "tables" (rather than "objects", which is just evidence of me thinking in terms of JSON).
I'm marking this resolved but please reopen if you think it's still unclear.

* How should dependencies be defined for projects which do not build
  distributions (non-package projects)?

In support of these two needs, there are two common solutions which are similar
to this proposal:

* ``requirements.txt`` files

* package `extras <https://packaging.python.org/en/latest/specifications/dependency-specifiers/#extras>`__

Both ``requirements.txt`` files and ``extras`` have limitations which this
standard seeks to overcome.

Note that the two use cases above describe two different types of projects
which this PEP seeks to support:

> * python packages, such as libraries

Reviewer: ```suggestion
* Python packages, such as libraries
```

* non-package projects, such as data science projects

Limitations of ``requirements.txt`` files
-----------------------------------------

Many projects may define one or more ``requirements.txt`` iles,
and may arrange them either at the project root (e.g. ``requirements.txt`` and
``test-requirements.txt``) or else in a directory (e.g.
``requirements/base.txt`` and ``requirements/test.txt``). However, there are
major issues with the use of requirements files in this way:

* There is no standardized naming convention such that tools can discover or
  use these files by name.

* ``requirements.txt`` files are *not standardized*, but instead provide
  options to ``pip``.

As a result, it is difficult to define tool behaviors based on
``requirements.txt`` files. They are not trivial to discover or identify by
name, and their contents may contain a mix of package specifiers and additional
``pip`` options.

> The lack of a standard for ``requirements.txt`` contents also means that not
> portable to any alternative tools which wish to process them other than

Reviewer: ```suggestion
The lack of a standard for ``requirements.txt`` contents also means they are
not portable to any alternative tools which wish to process them other than
```

``pip``.

Additionally, ``requirements.txt`` files require a file per dependency list.
For some use-cases, this makes the marginal cost of dependency groupings high,
relative to their benefit.
A terser declaration is beneficial to projects with a number of small groups of
dependencies.

In contrast with this, Dependency Groups are defined at a well known location
in ``pyproject.toml`` with fully standardized contents. Not only will they have
immediate utility, but they will also serve as a starting point for future
standards.

Limitations of ``extras``
-------------------------

``extras`` are additional package metadata declared in the
``[project.optional-dependencies]`` table. They provide names for lists of
package specifiers which are published as part of a package's metadata, and
which a user can request under that name, as in ``pip install 'foo[bar]'`` to
install ``foo`` with the ``bar`` extra.

Because ``extras`` are package metadata, they are not usable when a project

> does not build a distribution (is not a package).

Reviewer: ```suggestion
does not build a distribution (i.e., is not a package).
```

  dependencies or the package itself

Use Cases
---------

The following use cases are considered important targets for this PEP. They are
defined in greater detail in the `Use Cases Appendix <use_cases>`_.

* Web Applications deployed via a non-python-packaging build process
* Libraries with unpublished dev dependency groups
* Data science projects with groups of dependencies but no core package
* *Input data* to lockfile generation (Dependency Groups should generally not
  be used as a location for locked dependency data)
* Input data to an environment manager, such as tox, Nox, or Hatch
* Embedded ``pyproject.toml`` data in scripts, as proposed in :pep:`723`
* Configurable IDE discovery of test and linter requirements

Regarding Poetry and PDM Dependency Groups
------------------------------------------

The existing Poetry and PDM tools already offer a feature which each calls
"Dependency Groups", but using non-standard data belonging to the ``poetry``
and ``pdm`` tools.
(PDM also uses extras for some Dependency Groups, and overlaps the notion
heavily with extras.)

This PEP does not support all of the features of Poetry and PDM, which, like

> ``requirements.txt`` files for ``pip``, support several nonstandard extensions

Reviewer: ```suggestion
``requirements.txt`` files for ``pip``, support several non-standard extensions
```

.. code:: toml

    [dependency-groups]
    test = ["pytest", "coverage"]
    docs = ["sphinx", "sphinx-rtd-theme"]
    typing = ["mypy", "types-requests"]
    typing-test = [{include = "typing"}, {include = "test"}, "useful-types"]

Note that none of these Dependency Group declarations implicitly install the
current package, its dependencies, or any optional dependencies.
Use of a Dependency Group like ``test`` to test a package requires that the
user's configuration or toolchain also installs ``.``. For example,

.. code-block:: shell

    $TOOL install-dependency-group test
    pip install -e .

> Could be used (supposing ``$TOOL`` is a tool which supports installing

Reviewer: ```suggestion
could be used (supposing ``$TOOL`` is a tool which supports installing
```

Dependency Groups) to build a testing environment.

This also allows for the ``docs`` dependency group to be used without
installing the project as a package:

.. code-block:: shell

    $TOOL install-dependency-group docs

Package Building
----------------

Build backends MUST NOT include Dependency Group data in built distributions as
package metadata.

It is valid to use Dependency Groups in the evaluation of dynamic metadata.
For example, a build backend may define ``dependencies`` as dynamic and use
dependency groups to compute the value of ``dependencies``.

For example, a build backend could define the following data to evaluate
equivalently to ``dependencies=["aiohttp", "sqlalchemy"]``:

.. code:: toml

>     [project]
>     dynamic = ["dependencies"]
>     [dependency-groups]
>     http = ["aiohttp"]
>     db = ["sqlalchemy"]
>     [tool.some-build-tool.dynamic]
>     dependencies = {
>         "dependency-groups": ["http", "db"]
>     }

Reviewer: ```suggestion
    [project]
    dynamic = ["dependencies"]
    
    [dependency-groups]
    http = ["aiohttp"]
    db = ["sqlalchemy"]
    
    [tool.some-build-tool.dynamic]
    dependencies = {
        "dependency-groups": ["http", "db"]
    }
```
